lib/operations/execute_operation.js

```diff
@@ -47,6 +47,8 @@ function executeOperation(topology, operation, callback) {
     } else if (operation.session.hasEnded) {
       throw new MongoError('Use of expired sessions is not permitted');
     }
+  } else if (operation.session && operation.session.explicit) {
```

Comment:
I think a little more work is required here. The goal of the change is to defer the work of determining session support until a server is selected, that should apply to implicit and explicit sessions - so the check a few lines above this for `hasSessionsSupport` still suffers the same issue as the originally reported issue: it's a synchronous check of an asynchronously updated state. 

The refactor might be a little more involved, here are some thoughts:
- can we remove the code to distinguish between `EXECUTE_WITH_SELECTION` and not? I think this work was done in the 4.0 branch, maybe its not possible in 3.6?
- the `executeWithServerSelection` will likely need to be pulled into `executeOperation`, because the session check should happen each time a server is selected (if a retry is required, for instance).

Reply:
Actually this issue only impacts the UnifiedTopology so this was the right spot to move this for the legacy topology I believe, but a few lines above (line 33) there's a `selectServerForSessionSupport` call that will be made. So I think yes I should move the check into the callback of that selection, but also that selection will fail if it can't find a server that supports sessions. So if we want to report this error properly for the unified Topology we'd need to see if we fail selection specifically because of no session support, does that make sense?

Reply:
Hm, if we believe that this issue is caused by synchronously checking for sessions support in `startSession` then this problem must exist for both unified and legacy topologies - perhaps it's just easier to surface in the unified because it is stricter about when and how operations are run.  I forgot we added that `selectServerForSessionSupport`, this would also seem to imply there's a bug there since that should theoretically prevent this issue from happening. I think you have the same problem here anyway, `shouldCheckForSessionSupport` is a sync check on data which may change as a result of async operations. 

What I'm proposing here is _always_ deferring the decision about sessions until the moment a server is selected for an operation. Once the server is selected, and a connection is checked out from the pool for that server, you are at a point-in-time where you can trust all the information at hand. It's here that you can say "did you provide a session? well sessions aren't supported on this connection, sorry", or "sessions are supported, but none was provided, so we should use an implicit session". The problem with the way `executeOperation` is written today is that we try to make that determination before any selection occurs at all, and thus might miss out on updated information gained _in the process of selecting a server_.